PT nightly optim failing both on cpu and gpu

```
FAILED test/prototype/test_low_bit_optim.py::TestOptim::test_optim_smoke_optim_name_AdamW8bit_bfloat16_device_cuda - torch._dynamo.exc.TorchRuntimeError: Failed running call_method lerp(*(OptimState8bit(signed=True, block_size=256, shape=(256, 32), device=cuda:0, requires_grad=False), FakeTensor(..., device='cuda:0', size=(256, 32), dtype=torch.bfloat16), 0.09999999999999998), **{}):
  expected dtype torch.float32 for `end`, but got dtype torch.bfloat16
  
  from user code:
     File "/opt/conda/envs/venv/lib/python3.9/site-packages/torchao/prototype/low_bit_optim/adam.py", line 136, in single_param_adam
      new_exp_avg = exp_avg.lerp(grad, 1 - beta1)
  
  Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
  
  
  You can suppress this exception and fall back to eager by setting:
      import torch._dynamo
      torch._dynamo.config.suppress_errors = True
  
  
  To execute this test, run the following from the base repo dir:
      python test/prototype/test_low_bit_optim.py TestOptim.test_optim_smoke_optim_name_AdamW8bit_bfloat16_device_cuda
  
  This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
  Traceback (most recent call last):
  FAILED test/prototype/test_low_bit_optim.py::TestOptim::test_optim_smoke_optim_name_AdamWFp8_bfloat16_device_cpu - torch._dynamo.exc.TorchRuntimeError: Failed running call_method lerp(*(OptimStateFp8(block_size=256, shape=(256, 32), device=cpu, requires_grad=False), FakeTensor(..., size=(256, 32), dtype=torch.bfloat16), 0.09999999999999998), **{}):
  expected dtype torch.float32 for `end`, but got dtype torch.bfloat16
  
  from user code:
    File "/home/ec2-user/actions-runner/_work/ao/ao/test-infra/.github/scripts/run_with_env_secrets.py", line 102, in <module>
      main()
    File "/home/ec2-user/actions-runner/_work/ao/ao/test-infra/.github/scripts/run_with_env_secrets.py", line 98, in main
      run_cmd_or_die(f"docker exec -t {container_name} /exec")
    File "/home/ec2-user/actions-runner/_work/ao/ao/test-infra/.github/scripts/run_with_env_secrets.py", line 39, in run_cmd_or_die
      raise RuntimeError(f"Command {cmd} failed with exit code {exit_code}")
  RuntimeError: Command docker exec -t e74d66a60a6291d2f59d9c62d0cc410d78f62234cd213495bb69f85b4691ff43 /exec failed with exit code 1
     File "/opt/conda/envs/venv/lib/python3.9/site-packages/torchao/prototype/low_bit_optim/adam.py", line 136, in single_param_adam
      new_exp_avg = exp_avg.lerp(grad, 1 - beta1)
  
  Set TORCH_LOGS="+dynamo" and TORCHDYNAMO_VERBOSE=1 for more information
  
  
  You can suppress this exception and fall back to eager by setting:
      import torch._dynamo
      torch._dynamo.config.suppress_errors = True
  
  
  To execute this test, run the following from the base repo dir:
      python test/prototype/test_low_bit_optim.py TestOptim.test_optim_smoke_optim_name_AdamWFp8_bfloat16_device_cpu
  
  This message can be suppressed by setting PYTORCH_PRINT_REPRO_ON_FAILURE=0
  === 10 failed, 1639 passed, 380 skipped, 123 warnings in 3083.48s (0:51:23) ====
  Error: Process completed with exit code 1.
```

--- a/.github/workflows/regression_test.yml
+++ b/.github/workflows/regression_test.yml
@@ -67,7 +67,7 @@ jobs:
 
     uses: pytorch/test-infra/.github/workflows/linux_job.yml@main
     with:
-      timeout: 60
+      timeout: 120
       runner: ${{ matrix.runs-on }}
       gpu-arch-type: ${{ matrix.gpu-arch-type }}
       gpu-arch-version: ${{ matrix.gpu-arch-version }}

--- a/torchao/prototype/low_bit_optim/adam.py
+++ b/torchao/prototype/low_bit_optim/adam.py
@@ -109,9 +109,6 @@ class _AdamBase(Optimizer):
 
 # this will work with any optim state tensor subclass that implements aten.lerp.Scalar and aten.copy_.default
 # and param tensor subclass that implements aten.add_.Tensor, and aten.addcdiv_.default
-# NOTE: right now all of our optimizer state subclasses will dequant to FP32, thus adam computation
-# will be done in FP32 (not purposely). we should explicitly cast all inputs to FP32 to ensure FP32
-# computation. will need to benchmark to ensure no slowdown.
 def single_param_adam(
     p: Tensor,
     grad: Tensor,
@@ -126,32 +123,36 @@ def single_param_adam(
     eps: float,
     is_adamw: bool,
 ):
+    # compute in FP32 for accurate calculations
+    p_f32 = p.float()
+    grad_f32 = grad.float()
+
     if not is_adamw:
-        grad = grad.add(p, alpha=weight_decay)
+        grad_f32 = grad_f32.add(p_f32, alpha=weight_decay)
 
     bias_correction1 = 1 - beta1**step
     bias_correction2 = 1 - beta2**step
 
     # keep high precision copy for param update
-    new_exp_avg = exp_avg.lerp(grad, 1 - beta1)
-    new_exp_avg_sq = exp_avg_sq.lerp(grad.square(), 1 - beta2)
+    exp_avg_f32 = exp_avg.float().lerp(grad_f32, 1 - beta1)
+    exp_avg_sq_f32 = exp_avg_sq.float().lerp(grad_f32.square(), 1 - beta2)
 
-    exp_avg.copy_(new_exp_avg)
-    exp_avg_sq.copy_(new_exp_avg_sq)
+    exp_avg.copy_(exp_avg_f32)
+    exp_avg_sq.copy_(exp_avg_sq_f32)
 
     if max_exp_avg_sq is not None:
-        new_max_exp_avg_sq = torch.maximum(max_exp_avg_sq, new_exp_avg_sq)
-        max_exp_avg_sq.copy_(new_max_exp_avg_sq)
-        denom = (new_max_exp_avg_sq.sqrt() / bias_correction2.sqrt()).add_(eps)
+        max_exp_avg_sq_f32 = torch.maximum(max_exp_avg_sq.float(), exp_avg_sq_f32)
+        max_exp_avg_sq.copy_(max_exp_avg_sq_f32)
+        denom = (max_exp_avg_sq_f32.sqrt() / bias_correction2.sqrt()).add_(eps)
     else:
-        denom = (new_exp_avg_sq.sqrt() / bias_correction2.sqrt()).add_(eps)
+        denom = (exp_avg_sq_f32.sqrt() / bias_correction2.sqrt()).add_(eps)
 
     step_size = lr / bias_correction1
     if is_adamw:
         # merge weight decay and param update in a single .add_() to make this work with quantized param
-        p.add_(-lr * weight_decay * p - step_size * new_exp_avg / denom)
+        p.add_(-lr * weight_decay * p_f32 - step_size * exp_avg_f32 / denom)
     else:
-        p.addcdiv_(new_exp_avg, denom, value=-step_size)
+        p.addcdiv_(exp_avg_f32, denom, value=-step_size)
 
 
 class Adam8bit(_AdamBase):

